src/ProgrammableDevice.cc

```diff
@@ -0,0 +1,109 @@
+#include "ProgrammableDevice.hh"
+
+#include "CommandController.hh"
+#include "MSXCliComm.hh"
+#include "MSXCPUInterface.hh"
+#include "MSXException.hh"
+#include "StringSetting.hh"
+#include "serialize.hh"
+
+#include "narrow.hh"
+
+namespace openmsx {
+
+ProgrammableDevice::ProgrammableDevice(const DeviceConfig& config)
+	: MSXDevice(config)
+	, portListSetting(
+		getCommandController(), tmpStrCat(getName(), " ports"),
+		"List of ports are added to the pool of IO ports this extension scans for", "")
+	, resetCallback(getCommandController(),
+		getName() + " reset callback",
+		"Tcl proc to call when reset was detected, no parameters and return values are "
+		"used", "", Setting::Save::NO)
+	, readCallback(getCommandController(),
+		getName() + " read callback",
+		"Defines a Tcl proc that receives a port number (byte) and returns a byte when "
+		"the MSX is reading a port from the device", "", Setting::Save::NO)
+	, writeCallback(getCommandController(),
+		getName() + " write callback",
+		"Defines a Tcl proc that receives a port number (byte) and a byte when the "
+		"MSX is writing a port from the device", "", Setting::Save::NO)
+{
+	portListSetting.attach(*this);
+	try {
+		outPorts = inPorts = parseAllPorts(portListSetting);
+	} catch (MSXException&) {
+		// silently ignore error in settings.xml
+		portListSetting.setString("");
+	}
+}
+
+ProgrammableDevice::~ProgrammableDevice()
+{
+	portListSetting.detach(*this);
+}
+
+IterableBitSet<256> ProgrammableDevice::parseAllPorts(const Setting& setting)
+{
+	Interpreter& interp = getCommandController().getInterpreter();
+	TclObject portStr(((StringSetting&)setting).getString());
+	IterableBitSet<256> ports = {};
+
+	for (unsigned n = 0; n < portStr.getListLength(interp); n++) {
+		int tmp = portStr.getListIndex(interp, n).getInt(interp);
+		if ((tmp < 0) || (tmp > 255)) {
+		        throw MSXException("outside range 0..255");
+		}
+		ports.set(narrow_cast<byte>(tmp));
+	}
+	return ports;
+}
+
+void ProgrammableDevice::update(const Setting& setting) noexcept
+{
+	try {
+		IterableBitSet<256> tmpPorts = parseAllPorts(setting);
+		unregisterPorts();
+		inPorts = outPorts = tmpPorts;
+		doRegisterPorts();
+	} catch (MSXException& e) {
+		getCliComm().printWarning(
+			"Parse error in \"",
+			portListSetting.getFullName(), "\": ", e.getMessage());
+	}
+}
+
+void ProgrammableDevice::writeIO(word port, byte value, EmuTime::param /*time*/)
+{
+	writeCallback.execute(port, value);
+}
+
+byte ProgrammableDevice::readIO(word port, EmuTime::param /* time */)
+{
+	try {
+		auto obj = readCallback.execute(port);
+		int tmp;
+		if (obj != TclObject()) {
```

Comment:
This is wrong. If this test is false, then we return an uninitialized `tmp` value. But shouldn't it be an error if `obj` is empty? Shouldn't the read-callback _always_ return a valid integer?